pkg/datapath/sockets/sockets.go

```diff
@@ -0,0 +1,294 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package sockets
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"net"
+	"syscall"
+
+	"github.com/sirupsen/logrus"
+	"github.com/vishvananda/netlink"
+	"github.com/vishvananda/netlink/nl"
+	"golang.org/x/sys/unix"
+
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+const (
+	sizeofSocketID      = 0x30
+	sizeofSocketRequest = sizeofSocketID + 0x8
+	sizeofSocket        = sizeofSocketID + 0x18
+	SOCK_DESTROY        = 21
+)
+
+var (
+	log          = logging.DefaultLogger.WithField(logfields.LogSubsys, "datapath-sockets")
+	native       = nl.NativeEndian()
+	networkOrder = binary.BigEndian
+)
+
+type SocketDestroyer interface {
+	Destroy(filter SocketFilter) error
+}
+
+type SocketFilter struct {
+	DestIp   net.IP
+	DestPort uint16
+	Family   uint8
+	Protocol uint8
+	// Optional callback function to determine whether a filtered socket needs to be destroyed
+	DestroyCB DestroySocketCB
+}
+
+type DestroySocketCB func(id netlink.SocketID) bool
+
+// Destroy destroys sockets matching the passed filter parameters using the
+// sock_diag netlink framework.
+//
+// Supported families in the filter: syscall.AF_INET, syscall.AF_INET6
+// Supported protocols in the filter: unix.IPPROTO_UDP
+func Destroy(filter SocketFilter) error {
+	family := filter.Family
+	protocol := filter.Protocol
+
+	if family != syscall.AF_INET && family != syscall.AF_INET6 {
+		return fmt.Errorf("unsupported family for socket destroy: %d", family)
+	}
+	var errs error
+	success, failed := 0, 0
+
+	// Query sockets matching the passed filter, and then destroy the filtered
+	// sockets.
+	switch protocol {
+	case unix.IPPROTO_UDP:
+		err := filterAndDestroyUDPSockets(family, func(sock netlink.SocketID, err error) {
+			if err != nil {
+				errs = errors.Join(errs, fmt.Errorf("UDP socket with filter [%v]: %w", filter, err))
+				failed++
+				return
+			}
+			if filter.MatchSocket(sock) {
+				log.Infof("socket %v", sock)
+				if err := destroySocket(sock, family, unix.IPPROTO_UDP); err != nil {
+					errs = errors.Join(errs, fmt.Errorf("destroying UDP socket with filter [%v]: %w", filter, err))
+					failed++
+					return
+				}
+				log.Debugf("Destroyed socket: %v", sock)
+				success++
+			}
+		})
+		if err != nil {
+			return fmt.Errorf("failed to get sockets with filter %v: %w", filter, err)
+		}
+
+	default:
+		return fmt.Errorf("unsupported protocol for socket destroy: %d", protocol)
+	}
+	if success > 0 || failed > 0 || errs != nil {
+		log.WithFields(logrus.Fields{
+			"filter":  filter,
+			"success": success,
+			"failed":  failed,
+			"errors":  errs,
+		}).Info("Forcefully terminated sockets")
+	}
+
+	return nil
+}
+
+func (f *SocketFilter) MatchSocket(socket netlink.SocketID) bool {
+	if socket.Destination.Equal(f.DestIp) && socket.DestinationPort == f.DestPort {
```

Comment:
We don't know whether a destination IP here is the one resolved from a ClusterIP or sent directly to a Pod IP. Is there a possibility of destroying sockets that we should not?

Reply:
Yes, that's precisely why we check the map.